daemon/oci_windows.go

```diff
@@ -344,6 +285,82 @@ func (daemon *Daemon) createSpecWindowsFields(c *container.Container, s *specs.S
 	return nil
 }
 
+var errInvalidCredentialSpecSecOpt = fmt.Errorf("invalid credential spec security option - value must be prefixed by 'file://', 'registry://', or 'base64://' followed by a non-empty value")
+
+// setWindowsCredentialSpec sets the spec's `Windows.CredentialSpec`
+// field if relevant
+func (daemon *Daemon) setWindowsCredentialSpec(c *container.Container, s *specs.Spec) error {
+	if c.HostConfig == nil || c.HostConfig.SecurityOpt == nil {
+		return nil
+	}
+
+	// TODO (jrouge/wk8): if provided with several security options, we silently ignore
+	// all but the last one (provided they're all valid, otherwise we do return an error);
+	// this doesn't seem like a great idea?
+	credentialSpec := ""
+
+	for _, secOpt := range c.HostConfig.SecurityOpt {
+		optSplits := strings.SplitN(secOpt, "=", 2)
+		if len(optSplits) != 2 {
+			return fmt.Errorf("invalid security option: no equals sign in supplied value %s", secOpt)
+		}
+		if !strings.EqualFold(optSplits[0], "credentialspec") {
+			return fmt.Errorf("security option not supported: %s", optSplits[0])
+		}
+
+		credSpecSplits := strings.SplitN(optSplits[1], "://", 2)
+		value := credSpecSplits[1]
+		if len(credSpecSplits) != 2 || credSpecSplits[1] == "" {
+			return errInvalidCredentialSpecSecOpt
+		}
+
+		var err error
+		switch strings.ToLower(credSpecSplits[0]) {
+		case "file":
+			if credentialSpec, err = readCredentialSpecFile(c.ID, daemon.root, filepath.Clean(value)); err != nil {
+				return err
+			}
+		case "registry":
+			if credentialSpec, err = readCredentialSpecRegistry(c.ID, value); err != nil {
+				return err
+			}
+		case "config":
```

Comment:
wondering (haven't fully thought this through); should we remove the `config` option, and make swarmkit convert it to `base64` (read the credential-spec from the config, and pass it as base64)?

@dperny @cpuguy83

Reply:
I had the same question, actually :)

Reply:
Yes, please ❤️

Reply:
Happy to make a PR for that on Swarmkit after this is merged is @dperny likes the idea too :)

Reply:
Does it require a swarmkit change? I think the node agent can just read the config and stick it into the the credential spec?

Reply:
@cpuguy83 : you may very well be right, I didn't look into it before opening my big mouth, much to my shame. I just assumed that it did require a swarmkit change.

Reply:
Would still very much like @dperny 's take on this though before I do make that change :)

Reply:
It would theoretically be a bit of a pain in the butt, but we could do that. You'd basically have to read the CredentialSpec at some point when converting to a container, and write it to base64. It should be done all in the engine, no swarmkit changes necessary.

Reply:
@wk8 just thinking of this; the credentials-spec file is always JSON, correct? Do we need to pass it as base64, or should we just pass as a plain string? (optionally, validating for well-formedness)

Reply:
Yes it is always JSON, so it could just be passed as a plain string.

In my experience though, that can fast become a pain with quote escaping when eg building a command line in a bash script, that's why I thought base64 might be a better idea? Happy to change though if you feel strongly about it.

JSON validation is an idea though. Didn't do it upfront because we don't seem to validate anything just yet - if we do validate, we should do it for all the different ways that exist; and that would arguably be somewhat of a breaking change. Happy to add it if you wish though.

Reply:
> In my experience though, that can fast become a pain with quote escaping when eg building a command line in a bash script, that's why I thought base64 might be a better idea

You mean when using (e.g.) `curl` to create API calls? Or when using a bash script that's using the (`docker`) cli?

If we want to add this as an option on the docker cli, I assume the cli itself would take care of properly encoding the JSON. (perhaps the option should be named `raw` in that case)

Should `stdin` be supported there? e.g.:

```
cat ./my-creds.json | docker service create --credential-specs=file://- ....
```

(using `-` as name for the file to indicate it's from `stdin`)

Reply:
Fair enough, changing to using raw strings.

Reply:
Can we go ahead and remove config from this?

Reply:
As we discussed on the maintainers' call yesterday, I'll do that in a follow-up PR.

Reply:
Why not just in a separate commit?

Reply:
Seems like the whole reason to add this is to remove config.

Reply:
Well, the original reason for this is to remove the whole awkward dance I had to do in the kubelet to write/clean up registry keys to add GMSA support there (see the PR message).

And I'll be happy to do that in a follow-up PR; it can't hurt to keep PRs to a reasonable size?

Reply:
Fine, I would prefer to have the change come in through a single PR but in separate commits but 🤷‍♂️

Reply:
It arguably is 2 different things; I'll make a follow-up PR as soon as this one is merged. Thanks!